sw/device/lib/testing/flash_ctrl_testutils.c

```diff
@@ -286,43 +287,69 @@ bool flash_ctrl_testutils_bank_erase(dif_flash_ctrl_state_t *flash_state,
   return retval;
 }
 
-uint32_t flash_ctrl_testutils_get_count(uint32_t *strike_counter) {
-  uint32_t addr = (uint32_t)strike_counter;
-  uint32_t val = abs_mmio_read32(addr);
-
-  for (size_t i = 0; i < 32; ++i) {
-    if (val >> i & 0x1) {
-      return i;
+enum {
+  kNonVolatileCounterFlashWords = 256,
+};
+static_assert(kNonVolatileCounterFlashWords ==
+                  kFlashCtrlTestUtilsCounterMaxCount,
+              "Word count must be equal to max count.");
+static_assert(
+    FLASH_CTRL_PARAM_BYTES_PER_WORD == sizeof(uint64_t),
+    "Elements of the counter array must be the same size as a flash word");
+extern char _non_volatile_counter_flash_words[];
+
+OT_SECTION(".non_volatile_counter_0")
+uint64_t nv_counter_0[kNonVolatileCounterFlashWords];
+OT_SECTION(".non_volatile_counter_1")
+uint64_t nv_counter_1[kNonVolatileCounterFlashWords];
+OT_SECTION(".non_volatile_counter_2")
+uint64_t nv_counter_2[kNonVolatileCounterFlashWords];
+
+static uint64_t *const kNvCounters[] = {
+    nv_counter_0,
+    nv_counter_1,
+    nv_counter_2,
+};
+
+uint32_t flash_ctrl_testutils_counter_get(size_t counter) {
+  CHECK(counter < ARRAYSIZE(kNvCounters));
+  CHECK((uint32_t)&_non_volatile_counter_flash_words ==
+        kNonVolatileCounterFlashWords);
+
+  // Use a reverse loop since `flash_ctrl_testutils_set_at_least()` can
+  // introduce gaps.
+  size_t i = kNonVolatileCounterFlashWords - 1;
+  for (; i < kNonVolatileCounterFlashWords; --i) {
+    if (kNvCounters[counter][i] == 0) {
+      break;
     }
   }
-  CHECK(false, "the counter is all zero?");
-  return -1;
+  return i + 1;
```

Comment:
Will the first one never get used? E.g. what's the expected behavior for the call `uint32_t event_idx = flash_ctrl_testutils_counter_get(0);`?

Reply:
Sorry, I don't follow. This loop starts from the end of the counter page (page = 2048 bytes) and breaks at the first zero flash word (i from 255 to 0). The value of the counter is `i + 1`, i.e. counter[255]=0 -> 256, no zeros -> 0 (i = UINT32_MAX + 1 = 0).